### DISCUSSION TITLE
Allow using Pool.connect() in a context manager

### DISCUSSION THREAD

**Discussion by leo-b** [sqlalchemy/sqlalchemy#11516] [Ideas]
I'd like to use pools created by create_pool_from_url() and create_async_pool_from_url() directly as documented [here ](https://docs.sqlalchemy.org/en/20/core/pooling.html#using-a-pool-instance-directly).

It would be nice if `pool.connect()` could be used in a context manager:
```
with pool.connect() as connection:
    # use the connection
    ...
```

Unfortunately the `_ConnectionFairy` proxied connection returned by `pool.connect()` currently does not support the context manager protocol.

**Comment by CaselIT** [comment]
Hi,

Any reason for not using an engine in this case?

**Comment by zzzeek** [comment]
hi -

we've backed away from promoting pool as a front-facing API object that can be used since it's an almost never used use case and supporting it is difficult.  a lot of the pool functions are dialect-specific and only come in when you use create_engine.

to that end, you can use connection fairy directly from the engine using `e.raw_connection()`.    we dont have the context manager hooks there at the moment.   but you can also get the raw connection from the sqlalchemy Connection and just do things normally:


```
with e.connect() as conn:
     raw_conn = conn.connection
```

this is documented at https://docs.sqlalchemy.org/en/20/core/connections.html#working-with-the-dbapi-cursor-directly

**Comment by leo-b** [comment]
OK, thanks. So I'll obtain my connections via the engine.
However I still have two questions regarding engine connections:

1. I have to use dialect specific _oracle+oracledb_ features. (E.g. `Connection.createlob()`). Those methods are available in the sync api via `conn.connection` because `_ConnectionFairy` delegates to the dbapi connection.
Unfortunately I cannot use `conn.connection` in an async environment because `AsyncConnection.connection` is not implemented. Using `conn.get_raw_connection()` provides an `AsyncAdapt_oracledb_connection` which interits from `AsyncAdapt_dbapi_connection`. But there is no delegation to the driver connection, so I guess I'll have to use `await conn.get_raw_connection().driver_connection`?

2) I am using fastapi to implement a microframework and I have to deal with database disconnects using the optimistic approach: If it turns out that a connection is disconnected during usage, the current request should fail and the connection and all connections of the pool should be marked for reconnect. I am using raw connections so I guess, this disconnect handling needs to be done in the context manager of Engine.connect(). However I couldn't find any code that catches disconnect exceptions in the `__exit__` handler of `engine.Connection`. Did I miss something?

**Reply by CaselIT** [reply]
1. yes, to get the actual oracle object you can use `(await conn.get_raw_connection()).driver_connection`. Your use case seems very niche, how come you are trying to use functions like createlob()? Sqlalchemy will use them when needed automatically. If you are interested in getting back those object you can disable the auto-read as mentioned here https://docs.sqlalchemy.org/en/20/dialects/oracle.html#lob-datatypes
2. In your case I would handle the disconnect on a "before_request_end" hook of the framework you are using. This way you can have that logic in a central place

**Reply by leo-b** [reply]
> how come you are trying to use functions like createlob()?

This was just an example. I need other methods like `queue()` and `gettype()` as well. And I'd like to use native cursors to be able to do multiple executes (using the same SQL with different bind params) on one cursor  and benefit from the lower overhead of using already prepared statements. (That's >20% faster execution times in my test case.)

>     2. In your case I would handle the disconnect on a "before_request_end" hook of the framework you are using. This way you can have that logic in a central place

But when I am using the dbapi connection (or the driver connection) directly, sqlalchemy won't notice that the connection got disconnected and it won't handle any disconnect-related exception raised when using it, right?
So when doing request-cleanup I'd have to
1. catch any unhandled database exception
2. check if it is a disconnect exception (`engine.dialect.is_disconnect()`?)
3. if yes, invalidate the connection
4. "close" (i.e. return) the connection (or exit the Engine.connect context manager)

Does that sound correct?
Maybe I could use `connection._handle_dbapi_exception()` for 2. and 3.?

Cheers,
--leo

**Reply by CaselIT** [reply]
Yes you would likely have to manually manage the exception.

Overall your description seems to be of an oracle custom application, maybe consider using oracledb directly? it does offer connection pooling https://python-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#connection-pooling, so I'm not sure what you gain by using sqlalchemy to manage the connections if you are using only raw oracledb connections
